nemo/collections/nlp/models/nlp_model.py

```diff
@@ -0,0 +1,89 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC
+
+import torch
+from megatron import mpu
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.overrides.data_parallel import LightningDistributedDataParallel
+
+from nemo.core.classes import ModelPT
+from nemo.utils import AppState, logging
+
+__all__ = ['NLPModel']
+
+
+class NLPModel(ModelPT, ABC):
+    def init_ddp_connection(self, global_rank: int, world_size: int, is_slurm_managing_tasks: bool = True) -> None:
+        """ Override LightningModule DDP initialization if using model parallel"""
+        app_state = AppState()
+
+        # we are able initialize megatron-lm model parallel and data parallel groups
+        # after initializing DDP with PTL.
+        if app_state.model_parallel_size is not None:
+            LightningModule.init_ddp_connection(self, global_rank, world_size, is_slurm_managing_tasks)
+            if app_state.model_parallel_group is None:
+                mpu.initialize_model_parallel(app_state.model_parallel_size)
+                app_state.model_parallel_group = mpu.get_model_parallel_group()
+                app_state.data_parallel_group = mpu.get_data_parallel_group()
+                app_state.model_parallel_rank = torch.distributed.get_rank(group=app_state.model_parallel_group)
+                app_state.data_parallel_rank = torch.distributed.get_rank(group=app_state.data_parallel_group)
+                logging.info(f'mp_rank: {app_state.model_parallel_rank}')
+                logging.info(f'dp_rank: {app_state.data_parallel_rank}')
+
```

Comment:
The else case returns some value, the if case doesnt? Maybe cache the result from the above `LightningModule.init_ddp_connection` and return it here?

Reply:
There's no return value. I removed `else` as it wasn't needed.